### PR TITLE
dvc: optimize building graph

### PR DESCRIPTION
**Pull request by Suor**
Use trie structure to efficiently check out, dep and stage paths being
prefixes of each other.

Will help with #2671.

### CODE REVIEW

**Comment by efiop** [comment]
@Suor Any benchmarks?

File: dvc/repo/__init__.py [hunk 1]
```diff
@@ -314,48 +314,57 @@ def _collect_graph(self, stages=None):
...
-                            str(outs[p]),
-                            outs[p].stage.relpath,
-                            str(out),
-                            out.stage.relpath,
-                        )
-                        raise OverlappingOutputPathsError(outs[p], out, msg)
+                # Check for overlapping outs
+                if outs.has_subtrie(out_key):
+                    parent = out
+                    overlapping = first(outs.values(prefix=out_key))
```
Comment (shcheklein) [inline dvc/repo/__init__.py]: just to double check - does it take into account situations like: `/a/b/test` in  `/a/b/testing`?

Comment (Suor) [reply dvc/repo/__init__.py]: Yes, it builds on `PathInfo(...).parts`.

**Comment by Suor** [comment]
Ok, waiting for second approval, unless there are more comments.

**Pull request merged**